File: fix_wav.py
"""
Fix voice recorder WAV files for Windows compatibility.

Many mini voice recorders produce WAV files in non-standard formats
(8-bit, ADPCM, unusual sample rates, bad headers) that Windows Media Player
and some apps cannot play. This script converts them to standard PCM 16-bit WAV,
which plays everywhere on Windows.

Usage:
  python fix_wav.py <file.wav> [<file2.wav> ...]
  python fix_wav.py <directory>   # processes all .wav in directory

Output: writes fixed files as <name>_fixed.wav in the same folder.
        Use --replace to overwrite originals instead.

If soundfile fails (rare formats), install ffmpeg and use:
  python compress_audio.py <file.wav> --format wav16
"""
from __future__ import annotations

import argparse
import shutil
import subprocess
import sys
from pathlib import Path

import soundfile as sf


def _fix_with_ffmpeg(input_path: Path, output_path: Path) -> bool:
    result = subprocess.run(
        [
            "ffmpeg", "-hide_banner", "-loglevel", "error",
            "-i", str(input_path), "-acodec", "pcm_s16le", "-y", str(output_path),
        ],
        capture_output=True,
        text=True,
    )
    if result.returncode != 0:
        print(f"  ffmpeg ERROR: {result.stderr.strip() or 'unknown'}")
        return False
    return True


def fix_wav(input_path: Path, output_path: Path) -> bool:
    """Convert WAV to standard PCM 16-bit. Returns True on success."""
    try:
        data, sample_rate = sf.read(str(input_path), dtype="int16")
        sf.write(str(output_path), data, sample_rate, subtype="PCM_16")
        return True
    except Exception as e:
        print(f"  soundfile failed: {e}")
        if output_path.exists():
            output_path.unlink()
        if shutil.which("ffmpeg"):
            print("  Trying ffmpeg fallback...")
            return _fix_with_ffmpeg(input_path, output_path)
        print("  Install ffmpeg and run: python compress_audio.py <file> --format wav16")
        return False


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Fix voice recorder WAV files for Windows playback.",
    )
    parser.add_argument(
        "paths",
        type=Path,
        nargs="+",
        help="WAV file(s) or directory to process.",
    )
    parser.add_argument(
        "--replace",
        action="store_true",
        help="Overwrite originals instead of creating *_fixed.wav",
    )
    args = parser.parse_args()

    wav_files: list[Path] = []
    for p in args.paths:
        if not p.exists():
            print(f"Error: '{p}' does not exist.")
            sys.exit(1)
        if p.is_file():
            if p.suffix.lower() == ".wav":
                wav_files.append(p)
            else:
                print(f"Skipping non-WAV: {p}")
        else:
            wav_files.extend(sorted(p.rglob("*.wav")))

    if not wav_files:
        print("No WAV files found.")
        sys.exit(0)

    print(f"Fixing {len(wav_files)} WAV file(s)...\n")
    ok = 0
    for wav in wav_files:
        if args.replace:
            out = wav.parent / (wav.stem + "_tmp.wav")
            success = fix_wav(wav, out)
            if success:
                out.replace(wav)
                ok += 1
                print(f"  OK: {wav.name}")
            else:
                out.unlink(missing_ok=True)
        else:
            out = wav.parent / (wav.stem + "_fixed.wav")
            success = fix_wav(wav, out)
            if success:
                ok += 1
                print(f"  OK: {wav.name} -> {out.name}")
    print(f"\nDone. Fixed {ok}/{len(wav_files)} files.")


if __name__ == "__main__":
    main()

File: fix_wav.mjs
/**
 * Fix voice recorder WAV files for Windows compatibility.
 *
 * Converts non-standard WAV formats to standard PCM 16-bit WAV
 * that plays everywhere on Windows. Requires ffmpeg on PATH.
 *
 * Usage:
 *   node fix_wav.mjs <file.wav> [<file2.wav> ...]
 *   node fix_wav.mjs <directory>
 *
 * Options:
 *   --replace   Overwrite originals instead of creating *_fixed.wav
 */

import { execFile } from "node:child_process"
import { existsSync, renameSync, unlinkSync } from "node:fs"
import { dirname, join } from "node:path"
import { readdirSync, statSync } from "node:fs"

function collectWavFiles(paths) {
  const wavFiles = []
  for (const p of paths) {
    if (!existsSync(p)) {
      console.error(`Error: '${p}' does not exist.`)
      process.exit(1)
    }
    const stat = statSync(p)
    if (stat.isFile()) {
      if (p.toLowerCase().endsWith(".wav")) wavFiles.push(p)
      else console.log(`Skipping non-WAV: ${p}`)
    } else {
      for (const entry of walkDir(p)) {
        if (entry.toLowerCase().endsWith(".wav")) wavFiles.push(entry)
      }
    }
  }
  return wavFiles.sort()
}

function* walkDir(dir) {
  for (const name of readdirSync(dir)) {
    const full = join(dir, name)
    if (statSync(full).isDirectory()) {
      yield* walkDir(full)
    } else {
      yield full
    }
  }
}

function fixWav(inputPath, outputPath) {
  return new Promise((resolve) => {
    execFile(
      "ffmpeg",
      ["-hide_banner", "-loglevel", "error", "-i", inputPath, "-acodec", "pcm_s16le", "-y", outputPath],
      (err, _stdout, stderr) => {
        if (err) {
          if (err.code === "ENOENT") {
            console.error("  ffmpeg not found. Install ffmpeg and add it to PATH.")
          } else {
            console.error(`  ffmpeg ERROR: ${stderr?.trim() || err.message}`)
          }
          resolve(false)
        } else {
          resolve(true)
        }
      }
    )
  })
}

async function main() {
  const args = process.argv.slice(2)
  const replaceIdx = args.indexOf("--replace")
  const replace = replaceIdx >= 0
  if (replace) args.splice(replaceIdx, 1)

  if (args.length === 0) {
    console.error("Usage: node fix_wav.mjs <file.wav> [<file2.wav> ...] [--replace]")
    process.exit(1)
  }

  const wavFiles = collectWavFiles(args)
  if (wavFiles.length === 0) {
    console.log("No WAV files found.")
    process.exit(0)
  }

  console.log(`Fixing ${wavFiles.length} WAV file(s)...\n`)
  let ok = 0

  for (const wav of wavFiles) {
    const dir = dirname(wav)
    const base = wav.replace(/\.wav$/i, "").replace(/^.*[/\\]/, "")
    const outPath = replace ? join(dir, base + "_tmp.wav") : join(dir, base + "_fixed.wav")

    const success = await fixWav(wav, outPath)

    if (success) {
      if (replace) {
        try {
          renameSync(outPath, wav)
          console.log(`  OK: ${base}.wav`)
        } catch (e) {
          console.error(`  ERROR renaming: ${e.message}`)
          unlinkSync(outPath)
        }
      } else {
        console.log(`  OK: ${base}.wav -> ${base}_fixed.wav`)
      }
      ok++
    }
  }

  console.log(`\nDone. Fixed ${ok}/${wavFiles.length} files.`)
}

main().catch((err) => {
  console.error(err)
  process.exit(1)
})
